File: results/dnd-rest-benchmark/lifecycle-runs/20260730T075409Z_pi_kimi-k2p7-code_ruby-stdlib/lib/campaigns.rb
# frozen_string_literal: true

require_relative 'config'
require_relative 'persistence'

# Campaigns: state, characters, and session logs.
module Campaigns
  def self.create(payload)
    data = validate_campaign(payload)
    return [:invalid] unless data

    Persistence.db do |d|
      existing = d.get_first_value('SELECT 1 FROM campaigns WHERE id = ?', data[:id])
      next [:conflict] if existing

      d.execute(
        'INSERT INTO campaigns (id, name, dm) VALUES (?, ?, ?)',
        [data[:id], data[:name], data[:dm]]
      )

      [:ok, {
        'id' => data[:id],
        'name' => data[:name],
        'dm' => data[:dm]
      }]
    end
  end

  def self.create_character(campaign_id, payload)
    return [:invalid] unless campaign_id.is_a?(String) && !campaign_id.empty?

    data = validate_character(payload)
    return [:invalid] unless data

    Persistence.db do |d|
      campaign_exists = d.get_first_value('SELECT 1 FROM campaigns WHERE id = ?', campaign_id)
      next [:not_found] unless campaign_exists

      existing = d.get_first_value('SELECT 1 FROM campaign_characters WHERE id = ?', data[:id])
      next [:conflict] if existing

      d.execute(
        'INSERT INTO campaign_characters (id, campaign_id, name, level, class) VALUES (?, ?, ?, ?, ?)',
        [data[:id], campaign_id, data[:name], data[:level], data[:class]]
      )

      [:ok, {
        'id' => data[:id],
        'name' => data[:name],
        'level' => data[:level],
        'class' => data[:class]
      }]
    end
  end

  def self.create_event(campaign_id, payload)
    return [:invalid] unless campaign_id.is_a?(String) && !campaign_id.empty?

    data = validate_event(payload)
    return [:invalid] unless data

    Persistence.db do |d|
      campaign_exists = d.get_first_value('SELECT 1 FROM campaigns WHERE id = ?', campaign_id)
      next [:not_found] unless campaign_exists

      existing = d.get_first_value('SELECT 1 FROM campaign_events WHERE id = ?', data[:id])
      next [:conflict] if existing

      d.execute(
        'INSERT INTO campaign_events (id, campaign_id, kind, summary) VALUES (?, ?, ?, ?)',
        [data[:id], campaign_id, data[:kind], data[:summary]]
      )

      [:ok, {
        'id' => data[:id],
        'kind' => data[:kind]
      }]
    end
  end

  def self.state(campaign_id)
    return [:invalid] unless campaign_id.is_a?(String) && !campaign_id.empty?

    Persistence.db do |d|
      row = d.get_first_row('SELECT name, dm FROM campaigns WHERE id = ?', campaign_id)
      next [:not_found] unless row

      name, dm = row

      characters = d.execute(
        'SELECT id, name, level, class FROM campaign_characters WHERE campaign_id = ? ORDER BY id',
        campaign_id
      ).map do |id, char_name, level, klass|
        { 'id' => id, 'name' => char_name, 'level' => level, 'class' => klass }
      end

      log_count = d.get_first_value('SELECT COUNT(*) FROM campaign_events WHERE campaign_id = ?', campaign_id)

      [:ok, {
        'id' => campaign_id,
        'name' => name,
        'dm' => dm,
        'characters' => characters,
        'log_count' => log_count
      }]
    end
  end

  def self.audit(campaign_id)
    return [:invalid] unless campaign_id.is_a?(String) && !campaign_id.empty?

    Persistence.db do |d|
      campaign_exists = d.get_first_value('SELECT 1 FROM campaigns WHERE id = ?', campaign_id)
      next [:not_found] unless campaign_exists

      events = d.get_first_value('SELECT COUNT(*) FROM campaign_events WHERE campaign_id = ?', campaign_id)
      quests = d.get_first_value('SELECT COUNT(*) FROM campaign_quests WHERE campaign_id = ?', campaign_id)
      npcs = d.get_first_value('SELECT COUNT(*) FROM campaign_npcs WHERE campaign_id = ?', campaign_id)
      sessions = d.get_first_value('SELECT COUNT(*) FROM campaign_sessions WHERE campaign_id = ?', campaign_id)

      [:ok, {
        'campaign_id' => campaign_id,
        'events' => events,
        'quests' => quests,
        'npcs' => npcs,
        'sessions' => sessions
      }]
    end
  end

  def self.export(campaign_id)
    return [:invalid] unless campaign_id.is_a?(String) && !campaign_id.empty?

    Persistence.db do |d|
      row = d.get_first_row('SELECT name FROM campaigns WHERE id = ?', campaign_id)
      next [:not_found] unless row

      name = row[0]
      characters = d.get_first_value('SELECT COUNT(*) FROM campaign_characters WHERE campaign_id = ?', campaign_id)
      quests = d.get_first_value('SELECT COUNT(*) FROM campaign_quests WHERE campaign_id = ?', campaign_id)
      npcs = d.get_first_value('SELECT COUNT(*) FROM campaign_npcs WHERE campaign_id = ?', campaign_id)
      inventory_items = d.get_first_value('SELECT COUNT(DISTINCT item_slug) FROM campaign_inventory WHERE campaign_id = ?', campaign_id)
      sessions = d.get_first_value('SELECT COUNT(*) FROM campaign_sessions WHERE campaign_id = ?', campaign_id)

      [:ok, {
        'campaign_id' => campaign_id,
        'name' => name,
        'characters' => characters,
        'quests' => quests,
        'npcs' => npcs,
        'inventory_items' => inventory_items,
        'sessions' => sessions,
        'schema_version' => Config::SCHEMA_VERSION
      }]
    end
  end

  def self.validate_campaign(payload)
    return nil unless payload.is_a?(Hash)

    id = payload['id']
    name = payload['name']
    dm = payload['dm']

    return nil unless id.is_a?(String) && !id.empty? &&
                      name.is_a?(String) && !name.empty? &&
                      dm.is_a?(String) && !dm.empty?

    { id: id, name: name, dm: dm }
  end
  private_class_method :validate_campaign

  def self.validate_character(payload)
    return nil unless payload.is_a?(Hash)

    id = payload['id']
    name = payload['name']
    level = payload['level']
    klass = payload['class']

    return nil unless id.is_a?(String) && !id.empty? &&
                      name.is_a?(String) && !name.empty? &&
                      level.is_a?(Integer) && level >= 1 && level <= 20 &&
                      klass.is_a?(String) && !klass.empty?

    { id: id, name: name, level: level, class: klass }
  end
  private_class_method :validate_character

  def self.validate_event(payload)
    return nil unless payload.is_a?(Hash)

    id = payload['id']
    kind = payload['kind']
    summary = payload['summary']

    return nil unless id.is_a?(String) && !id.empty? &&
                      kind.is_a?(String) && !kind.empty? &&
                      (summary.nil? || summary.is_a?(String))

    { id: id, kind: kind, summary: summary }
  end
  private_class_method :validate_event
end

File: results/dnd-rest-benchmark/lifecycle-runs/20260730T150315Z_pi_kimi-k2p7-code_rust-stdlib/src/campaigns.rs
use crate::json::{
    escape_json, extract_int, extract_objects, extract_string, extract_string_array,
    extract_top_array_content,
};
use crate::store::{
    campaign_exists, character_exists, event_exists, session_create, session_exists, session_get_next,
    session_set_attendance, sqlite_exec, sqlite_query,
};

pub enum CampaignError {
    BadRequest,
    NotFound,
    Conflict,
}

/// Parse a campaign id from a path like `/v1/campaigns/{id}/suffix`.
pub fn parse_campaign_path<'a>(path: &'a str, suffix: &'a str) -> Option<&'a str> {
    let rest = path.strip_prefix("/v1/campaigns/")?;
    let id = rest.strip_suffix(suffix)?;
    if id.is_empty() {
        return None;
    }
    Some(id)
}

/// Create a new campaign.
///
/// All fields (`id`, `name`, `dm`) are required and must be non-empty strings.
pub fn create_campaign(body: &str) -> Result<String, CampaignError> {
    let id = extract_string(body, "id").ok_or(CampaignError::BadRequest)?;
    let name = extract_string(body, "name").ok_or(CampaignError::BadRequest)?;
    let dm = extract_string(body, "dm").ok_or(CampaignError::BadRequest)?;
    if id.is_empty() || name.is_empty() || dm.is_empty() {
        return Err(CampaignError::BadRequest);
    }
    if campaign_exists(id).map_err(|_| CampaignError::BadRequest)? {
        return Err(CampaignError::Conflict);
    }
    let sql = format!(
        "INSERT INTO campaigns (id, name, dm) VALUES ('{}', '{}', '{}');",
        crate::store::sql_escape(id),
        crate::store::sql_escape(name),
        crate::store::sql_escape(dm)
    );
    sqlite_exec(&sql).map_err(|_| CampaignError::BadRequest)?;
    Ok(format!(
        r#"{{"id":"{}","name":"{}","dm":"{}"}}"#,
        escape_json(id),
        escape_json(name),
        escape_json(dm)
    ))
}

/// Add a player character to a campaign.
///
/// Requires a valid campaign, non-empty `id`/`name`/`class`, and a level of at
/// least 1.
pub fn add_character(campaign_id: &str, body: &str) -> Result<String, CampaignError> {
    if !campaign_exists(campaign_id).map_err(|_| CampaignError::BadRequest)? {
        return Err(CampaignError::NotFound);
    }
    let id = extract_string(body, "id").ok_or(CampaignError::BadRequest)?;
    let name = extract_string(body, "name").ok_or(CampaignError::BadRequest)?;
    let level = extract_int(body, "level").ok_or(CampaignError::BadRequest)?;
    let class = extract_string(body, "class").ok_or(CampaignError::BadRequest)?;
    if id.is_empty() || name.is_empty() || class.is_empty() || level < 1 {
        return Err(CampaignError::BadRequest);
    }
    if character_exists(campaign_id, id).map_err(|_| CampaignError::BadRequest)? {
        return Err(CampaignError::Conflict);
    }
    let sql = format!(
        "INSERT INTO characters (campaign_id, id, name, level, class) VALUES ('{}', '{}', '{}', {}, '{}');",
        crate::store::sql_escape(campaign_id),
        crate::store::sql_escape(id),
        crate::store::sql_escape(name),
        level,
        crate::store::sql_escape(class)
    );
    sqlite_exec(&sql).map_err(|_| CampaignError::BadRequest)?;
    Ok(format!(
        r#"{{"id":"{}","name":"{}","level":{},"class":"{}"}}"#,
        escape_json(id),
        escape_json(name),
        level,
        escape_json(class)
    ))
}

/// Add a log event to a campaign.
///
/// `id` and `kind` must be non-empty. `summary` is required but may be empty.
pub fn add_event(campaign_id: &str, body: &str) -> Result<String, CampaignError> {
    if !campaign_exists(campaign_id).map_err(|_| CampaignError::BadRequest)? {
        return Err(CampaignError::NotFound);
    }
    let id = extract_string(body, "id").ok_or(CampaignError::BadRequest)?;
    let kind = extract_string(body, "kind").ok_or(CampaignError::BadRequest)?;
    let summary = extract_string(body, "summary").ok_or(CampaignError::BadRequest)?;
    if id.is_empty() || kind.is_empty() {
        return Err(CampaignError::BadRequest);
    }
    if event_exists(campaign_id, id).map_err(|_| CampaignError::BadRequest)? {
        return Err(CampaignError::Conflict);
    }
    let sql = format!(
        "INSERT INTO events (campaign_id, id, kind, summary) VALUES ('{}', '{}', '{}', '{}');",
        crate::store::sql_escape(campaign_id),
        crate::store::sql_escape(id),
        crate::store::sql_escape(kind),
        crate::store::sql_escape(summary)
    );
    sqlite_exec(&sql).map_err(|_| CampaignError::BadRequest)?;
    Ok(format!(
        r#"{{"id":"{}","kind":"{}"}}"#,
        escape_json(id),
        escape_json(kind)
    ))
}

/// Read a campaign summary including its characters and event count.
pub fn read_campaign_state(campaign_id: &str) -> Result<String, CampaignError> {
    if !campaign_exists(campaign_id).map_err(|_| CampaignError::BadRequest)? {
        return Err(CampaignError::NotFound);
    }
    let sql = format!(
        "SELECT id, name, dm FROM campaigns WHERE id = '{}';",
        crate::store::sql_escape(campaign_id)
    );
    let output = sqlite_query(&sql).map_err(|_| CampaignError::BadRequest)?;
    let content = extract_top_array_content(&output).ok_or(CampaignError::NotFound)?;
    let objects = extract_objects(content);
    if objects.is_empty() {
        return Err(CampaignError::NotFound);
    }
    let obj = objects[0];
    let id = extract_string(obj, "id").ok_or(CampaignError::BadRequest)?;
    let name = extract_string(obj, "name").ok_or(CampaignError::BadRequest)?;
    let dm = extract_string(obj, "dm").ok_or(CampaignError::BadRequest)?;

    let char_sql = format!(
        "SELECT id, name, level, class FROM characters WHERE campaign_id = '{}' ORDER BY rowid;",
        crate::store::sql_escape(campaign_id)
    );
    let char_output = sqlite_query(&char_sql).map_err(|_| CampaignError::BadRequest)?;
    let char_content = extract_top_array_content(&char_output).unwrap_or("");
    let mut chars_json = Vec::new();
    for c_obj in extract_objects(char_content) {
        let cid = extract_string(c_obj, "id").ok_or(CampaignError::BadRequest)?;
        let cname = extract_string(c_obj, "name").ok_or(CampaignError::BadRequest)?;
        let level = extract_int(c_obj, "level").ok_or(CampaignError::BadRequest)?;
        let class = extract_string(c_obj, "class").ok_or(CampaignError::BadRequest)?;
        chars_json.push(format!(
            r#"{{"id":"{}","name":"{}","level":{},"class":"{}"}}"#,
            escape_json(cid),
            escape_json(cname),
            level,
            escape_json(class)
        ));
    }

    let count_sql = format!(
        "SELECT COUNT(*) AS log_count FROM events WHERE campaign_id = '{}';",
        crate::store::sql_escape(campaign_id)
    );
    let count_output = sqlite_query(&count_sql).map_err(|_| CampaignError::BadRequest)?;
    let count_content = extract_top_array_content(&count_output).unwrap_or("");
    let count_objects = extract_objects(count_content);
    let log_count = if count_objects.is_empty() {
        0
    } else {
        extract_int(count_objects[0], "log_count").unwrap_or(0)
    };

    Ok(format!(
        r#"{{"id":"{}","name":"{}","dm":"{}","characters":[{}],"log_count":{}}}"#,
        escape_json(id),
        escape_json(name),
        escape_json(dm),
        chars_json.join(","),
        log_count
    ))
}

/// Parse a campaign session attendance path:
/// `/v1/campaigns/{id}/sessions/{session_id}/attendance`.
pub fn parse_campaign_session_path<'a>(path: &'a str) -> Option<(&'a str, &'a str)> {
    let rest = path.strip_prefix("/v1/campaigns/")?;
    let marker = "/sessions/";
    let mid = rest.find(marker)?;
    let campaign_id = &rest[..mid];
    if campaign_id.is_empty() {
        return None;
    }
    let after = &rest[mid + marker.len()..];
    let session_id = after.strip_suffix("/attendance")?;
    if session_id.is_empty() {
        return None;
    }
    Some((campaign_id, session_id))
}

/// Schedule a new campaign session.
///
/// POST /v1/campaigns/{id}/sessions
pub fn schedule_session(campaign_id: &str, body: &str) -> Result<String, CampaignError> {
    if !campaign_exists(campaign_id).map_err(|_| CampaignError::BadRequest)? {
        return Err(CampaignError::NotFound);
    }
    let id = extract_string(body, "id").ok_or(CampaignError::BadRequest)?;
    let starts_at = extract_string(body, "starts_at").ok_or(CampaignError::BadRequest)?;
    let duration_minutes = extract_int(body, "duration_minutes").ok_or(CampaignError::BadRequest)?;
    let agenda = extract_string_array(body, "agenda").ok_or(CampaignError::BadRequest)?;
    if id.is_empty() || starts_at.is_empty() || duration_minutes < 1 {
        return Err(CampaignError::BadRequest);
    }
    if session_exists(campaign_id, id).map_err(|_| CampaignError::BadRequest)? {
        return Err(CampaignError::Conflict);
    }
    session_create(campaign_id, id, starts_at, duration_minutes, &agenda)
        .map_err(|_| CampaignError::BadRequest)?;
    Ok(format!(
        r#"{{"id":"{}","starts_at":"{}","duration_minutes":{},"agenda_count":{}}}"#,
        escape_json(id),
        escape_json(starts_at),
        duration_minutes,
        agenda.len()
    ))
}

/// Record attendance for a specific campaign session.
///
/// POST /v1/campaigns/{id}/sessions/{session_id}/attendance
pub fn record_attendance(
    campaign_id: &str,
    session_id: &str,
    body: &str,
) -> Result<String, CampaignError> {
    if !campaign_exists(campaign_id).map_err(|_| CampaignError::BadRequest)? {
        return Err(CampaignError::NotFound);
    }
    if !session_exists(campaign_id, session_id).map_err(|_| CampaignError::BadRequest)? {
        return Err(CampaignError::NotFound);
    }
    let present = extract_string_array(body, "present").ok_or(CampaignError::BadRequest)?;
    let absent = extract_string_array(body, "absent").ok_or(CampaignError::BadRequest)?;
    session_set_attendance(campaign_id, session_id, &present, &absent)
        .map_err(|_| CampaignError::BadRequest)?;
    Ok(format!(
        r#"{{"session_id":"{}","present_count":{},"absent_count":{}}}"#,
        escape_json(session_id),
        present.len(),
        absent.len()
    ))
}

/// Return a deterministic audit summary for a campaign.
///
/// GET /v1/campaigns/{id}/audit
pub fn audit_campaign(campaign_id: &str) -> Result<String, CampaignError> {
    if !campaign_exists(campaign_id).map_err(|_| CampaignError::BadRequest)? {
        return Err(CampaignError::NotFound);
    }
    let events = count_rows(campaign_id, "events", "1=1")?;
    let quests = count_rows(campaign_id, "quests", "1=1")?;
    let npcs = count_rows(campaign_id, "npcs", "1=1")?;
    let sessions = count_rows(campaign_id, "campaign_sessions", "1=1")?;
    Ok(format!(
        r#"{{"campaign_id":"{}","events":{},"quests":{},"npcs":{},"sessions":{}}}"#,
        escape_json(campaign_id),
        events,
        quests,
        npcs,
        sessions
    ))
}

/// Return a deterministic export summary for a campaign.
///
/// GET /v1/campaigns/{id}/export
pub fn export_campaign(campaign_id: &str) -> Result<String, CampaignError> {
    if !campaign_exists(campaign_id).map_err(|_| CampaignError::BadRequest)? {
        return Err(CampaignError::NotFound);
    }
    let sql = format!(
        "SELECT id, name FROM campaigns WHERE id = '{}';",
        crate::store::sql_escape(campaign_id)
    );
    let output = sqlite_query(&sql).map_err(|_| CampaignError::BadRequest)?;
    let content = extract_top_array_content(&output).ok_or(CampaignError::NotFound)?;
    let objects = extract_objects(content);
    if objects.is_empty() {
        return Err(CampaignError::NotFound);
    }
    let name = extract_string(objects[0], "name").ok_or(CampaignError::BadRequest)?;

    let characters = count_rows(campaign_id, "characters", "1=1")?;
    let quests = count_rows(campaign_id, "quests", "1=1")?;
    let npcs = count_rows(campaign_id, "npcs", "1=1")?;
    let inventory_items = count_inventory_rows(campaign_id)?;
    let sessions = count_rows(campaign_id, "campaign_sessions", "1=1")?;

    Ok(format!(
        r#"{{"campaign_id":"{}","name":"{}","characters":{},"quests":{},"npcs":{},"inventory_items":{},"sessions":{},"schema_version":1}}"#,
        escape_json(campaign_id),
        escape_json(name),
        characters,
        quests,
        npcs,
        inventory_items,
        sessions
    ))
}

fn count_rows(campaign_id: &str, table: &str, predicate: &str) -> Result<i64, CampaignError> {
    let sql = format!(
        "SELECT COUNT(*) AS cnt FROM {} WHERE campaign_id = '{}' AND {};",
        table,
        crate::store::sql_escape(campaign_id),
        predicate
    );
    let output = sqlite_query(&sql).map_err(|_| CampaignError::BadRequest)?;
    let content = extract_top_array_content(&output).unwrap_or("");
    let objects = extract_objects(content);
    if objects.is_empty() {
        return Ok(0);
    }
    extract_int(objects[0], "cnt").ok_or(CampaignError::BadRequest)
}

fn count_inventory_rows(campaign_id: &str) -> Result<i64, CampaignError> {
    let sql = format!(
        "SELECT COUNT(*) AS cnt FROM (SELECT item_slug FROM campaign_inventory WHERE campaign_id = '{}' AND quantity > 0 UNION SELECT item_slug FROM character_equipment WHERE campaign_id = '{}' AND quantity > 0);",
        crate::store::sql_escape(campaign_id),
        crate::store::sql_escape(campaign_id)
    );
    let output = sqlite_query(&sql).map_err(|_| CampaignError::BadRequest)?;
    let content = extract_top_array_content(&output).unwrap_or("");
    let objects = extract_objects(content);
    if objects.is_empty() {
        return Ok(0);
    }
    extract_int(objects[0], "cnt").ok_or(CampaignError::BadRequest)
}

/// Return the next scheduled campaign session.
///
/// GET /v1/campaigns/{id}/sessions/next
pub fn next_session(campaign_id: &str) -> Result<String, CampaignError> {
    if !campaign_exists(campaign_id).map_err(|_| CampaignError::BadRequest)? {
        return Err(CampaignError::NotFound);
    }
    match session_get_next(campaign_id).map_err(|_| CampaignError::BadRequest)? {
        Some((id, starts_at, _duration, agenda_count)) => Ok(format!(
            r#"{{"id":"{}","starts_at":"{}","agenda_count":{}}}"#,
            escape_json(&id),
            escape_json(&starts_at),
            agenda_count
        )),
        None => Err(CampaignError::NotFound),
    }
}
